### PR TITLE
Add Plot to Deephaven V2

### LINKED ISSUE #1965: Refactor the Plot module for Deephaven V2
to create a curated list of many existing overloads in the module that are more suitable for use in Python. 

part of #1263

### PR DESCRIPTION
**Pull request by jmao-denver**
WIP still, but is ready for some testing on the functionality that we care most

Fixes #1965

### CODE REVIEW

File: pyintegration/deephaven2/plot/figure.py [hunk 1]
```diff
@@ -100,15 +100,15 @@ def axes(
         name: str = None,
         axes: int = None,
         remove_series: List[str] = None,
-        plot_style: Union[str,PlotStyle] = None,
+        plot_style: Union[str, PlotStyle] = None,
     ) -> Figure:
         """TODO pydoc
```
Comment (jjbrosnan) [inline pyintegration/deephaven2/plot/figure.py:105]: Looks like the pydoc comments are mostly (if not entirely) done for this.  Comment still necessary?

File: pyintegration/deephaven2/plot/figure.py [hunk 2]
```diff
@@ -135,22 +135,22 @@ def axes(
         j_figure = self.j_figure
 
         if {"axes"}.issubset(non_null_args):
```
Comment (jjbrosnan) [inline pyintegration/deephaven2/plot/figure.py:137]: Is there an advantage to doing it this way as opposed to:

```suggestion
        if "axes" in non_null_args:
```

File: pyintegration/deephaven2/plot/figure.py [hunk 3]
```diff
@@ -162,42 +162,42 @@ def axes(
...
+        min: Union[str, float] = None,
+        max: Union[str, float] = None,
         invert: bool = None,
         log: bool = None,
         business_time: bool = None,
-        calendar: Union[str,BusinessCalendar] = None,
+        calendar: Union[str, BusinessCalendar] = None,
         transform: AxisTransform = None,
     ) -> Figure:
         """TODO pydoc
```
Comment (jjbrosnan) [inline pyintegration/deephaven2/plot/figure.py:182]: Same comment as before.  This looks complete for pydoc.  Is something still missing or is this done?

File: pyintegration/deephaven2/plot/figure.py [hunk 4]
```diff
@@ -552,22 +552,22 @@ def chart_legend(
...
             non_null_args = non_null_args.difference({"color"})
             f_called = True
 
         if {"font"}.issubset(non_null_args):
-            j_figure = self.j_figure.legendFont(font)
+            j_figure = j_figure.legendFont(font)
             non_null_args = non_null_args.difference({"font"})
             f_called = True
 
         if {"font_family", "font_style", "font_size"}.issubset(non_null_args):
```
Comment (jjbrosnan) [inline pyintegration/deephaven2/plot/figure.py:564]: Ok, I see here that my previous question about syntax wouldn't work as well here.  I'm still not sure how I feel about this syntax, though.  There are more "pythonic" ways to check for subsets in sets.  In particular, for font, font style, and font size, there may be a better way to handle it in a single if block rather than multiple.

File: pyintegration/deephaven2/plot/figure.py (hunk 1 above)
Comment (jmao-denver) [reply pyintegration/deephaven2/plot/figure.py]: yes, still missing the info on what the function does.

File: pyintegration/deephaven2/plot/figure.py (hunk 4 above)
Comment (jjbrosnan) [reply pyintegration/deephaven2/plot/figure.py:568]: I'm closing this comment.  This can be revisited later on.

**Review by hythloda — DISMISSED** [review]

**Review by jjbrosnan — APPROVED** [review]

File: pyintegration/deephaven2/plot/figure.py (hunk 2 above)
Comment (chipkent) [reply pyintegration/deephaven2/plot/figure.py:141]: The code generation is more simple with the existing case since it properly handles multi-argument cases.  `in` may be faster, but I doubt it makes a practical difference for these use cases, where the plot rendering time is the slow part.  I'd just delay any change like that until we can justify more complexity for a needed speed improvement.

File: Generators/src/main/java/io/deephaven/pythonPreambles/plotV2.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+#
+#   Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+""" This module implements the Figure class for creating plots. """
```
Comment (chipkent) [inline Generators/src/main/java/io/deephaven/pythonPreambles/plotV2.py:4]: pydoc

File: Generators/src/main/java/io/deephaven/pythonPreambles/plotV2.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            return obj.value.j_object
+        else:
+            return obj.value
+    elif isinstance(obj, Sequence):
+        import numpy
+        np_array = numpy.array(obj)
+        dtype = dtypes.from_np_dtype(np_array.dtype)
+        return dtypes.array(dtype, np_array)
+    elif isinstance(obj, Callable):
+        # TODO: support callables
```
Comment (chipkent) [inline Generators/src/main/java/io/deephaven/pythonPreambles/plotV2.py:78]: This should be supported.  I think it just needs to wrap the callable in a `PythonFunction`.
https://github.com/deephaven/deephaven-core/blob/5c1b52c85ab66a3fb77a0127095453276d024fe2/Integrations/src/main/java/io/deephaven/integrations/python/PythonFunction.java

File: pyintegration/deephaven2/plot/figure.py [hunk 7]
```diff
@@ -0,0 +1,2829 @@
...
+    def axis(
+        self,
+        dim: int = None,
+        t: Union[Table, SelectableDataSet] = None,
+        label: str = None,
+        color: Union[str, int, Color] = None,
+        font: Font = None,
+        font_family: str = None,
+        font_size: int = None,
+        font_style: str = None,
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/figure.py:175]: I thought you were consolidating all of this down to `font`.

File: pyintegration/deephaven2/plot/figure.py [hunk 8]
```diff
@@ -0,0 +1,2829 @@
...
+        if non_null_args == {"series_name", "t", "category", "y", "by"}:
+            return Figure(self.j_figure.catPlotBy(series_name, t, category, y, by))
+        if non_null_args == {"series_name", "category", "y", "y_low", "y_high"}:
+            return Figure(self.j_figure.catErrorBar(series_name, category, y, y_low, y_high))
+        elif non_null_args == {"series_name", "t", "category", "y", "y_low", "y_high"}:
+            return Figure(self.j_figure.catErrorBar(series_name, t, category, y, y_low, y_high))
+        if non_null_args == {"series_name", "t", "category", "y", "y_low", "y_high", "by"}:
+            return Figure(self.j_figure.catErrorBarBy(series_name, t, category, y, y_low, y_high, by))
+        else:
+            raise DHError(f"unsupported parameter combination: {non_null_args}")
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/figure.py:1116]: if vs elif looks to be misgenerated for this case.

File: pyintegration/deephaven2/plot/figure.py [hunk 9]
```diff
@@ -0,0 +1,2829 @@
...
+            by = _convert_j("by", by, [List[str]])
+
+        if non_null_args == {"series_name", "x", "open", "high", "low", "close"}:
+            return Figure(self.j_figure.ohlcPlot(series_name, x, open, high, low, close))
+        elif non_null_args == {"series_name", "t", "x", "open", "high", "low", "close"}:
+            return Figure(self.j_figure.ohlcPlot(series_name, t, x, open, high, low, close))
+        if non_null_args == {"series_name", "t", "x", "open", "high", "low", "close", "by"}:
+            return Figure(self.j_figure.ohlcPlotBy(series_name, t, x, open, high, low, close, by))
+        else:
+            raise DHError(f"unsupported parameter combination: {non_null_args}")
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/figure.py:1223]: misgenerated here as well

File: pyintegration/deephaven2/plot/figure.py [hunk 10]
```diff
@@ -0,0 +1,2829 @@
...
+    elif isinstance(obj, Enum):
+        if isinstance(obj.value, JObjectWrapper):
+            return obj.value.j_object
+        else:
+            return obj.value
+    elif isinstance(obj, Sequence):
+        import numpy
+        np_array = numpy.array(obj)
+        dtype = dtypes.from_np_dtype(np_array.dtype)
+        return dtypes.array(dtype, np_array)
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/figure.py:78]: 1) move numpy import to the top?
2) it looks easy to add numpy type support.  We either need to do it or have a ticket.  If we do it, the generated type annotations also need to change.  I'm ok with a ticket to get into next release.

File: pyintegration/tests/test_plot/test_figure.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        self.test_table = None
+
+    def test_figure(self):
+        figure = Figure()
+        new_f = figure.plot_xy("plot1", self.test_table, x="a", y="b")
+        plot1 = new_f.show()
+        new_f = figure.plot_xy("plot2", x=[1, 2, 3], y=[1.0, 2.0, 3.0])
+        plot2 = new_f.show()
+        self.assertIsNotNone(new_f)
+
```
Comment (chipkent) [inline pyintegration/tests/test_plot/test_figure.py:25]: @jjbrosnan @hythloda Can you help @jmao-denver create more comprehensive tests?  Coverage is low.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: Generators/Generators.gradle [hunk 12]
```diff
@@ -314,6 +328,10 @@ tasks.register 'generateAll', {
     t.description = 'Runs all code generators'
 }
 
+tasks.register 'generatePythonPlot2', {
+    Task t -> t.dependsOn ':Generators:generatePythonFigureWrapper2'
+}
```
[the remark this answers is not in the document]
Comment (niloc132) [reply Generators/Generators.gradle:333]: The syntax is weird, but all this does is create a task that basically calls another task. Since we're actually in :Generators already, it may not be necessary to have this at all - it basically is just an alias for generatePythonFigureWrapper2

File: Generators/Generators.gradle [hunk 13]
```diff
@@ -292,6 +292,20 @@ task generatePythonFigureWrapperAssert(type: JavaExec, dependsOn: 'classes') {
...
+
+    main = 'io.deephaven.plot.util.GeneratePyV2FigureAPI'
+    args devRoot, 'false'
+    classpath = sourceSets.main.runtimeClasspath
+    workingDir = workDir
+    systemProperty 'workspace', workspace
+    systemProperty 'devroot', devRoot
+    systemProperty 'Configuration.rootFile', 'dh-defaults.prop'
+    dependsOn generatePyDoc, ':Generators:generateFigureImmutable'
+}
```
[the remark this answers is not in the document]
Comment (niloc132) [reply Generators/Generators.gradle:307]: Personal opinion: it is better to have the generator run every time than mess with the assert, since the assert effectively does the same work, then blames you for not having run the generator, which _would have done the same thing without the finger wag_. That said, I'd go further and never check in the generated output unless we absolutely have to, and then always run the generate task as part of the build.

That said, this is an exec task, has no inputs or outputs declared, and yes, nothing calls it, which is going to be a problem. If the output is checked in, I could see making an assert task (and making `check` depend on that).

File: Plot/src/main/java/io/deephaven/plot/Axes.java [hunk 14]
```diff
@@ -1023,62 +1054,62 @@ OHLCDataSeries ohlcPlot(final Comparable seriesName, final IndexableNumericData
...
+     * @param time column in {@code t} that holds the time data
+     * @param open column in {@code t} that holds the open data
+     * @param high column in {@code t} that holds the high data
+     * @param low column in {@code t} that holds the low data
+     * @param close column in {@code t} that holds the close data
      * @return dataset created for plot
      */
-    OHLCDataSeries ohlcPlot(final Comparable seriesName, final Table t, final String timeCol, final String openCol,
-            final String highCol, final String lowCol, final String closeCol);
+    OHLCDataSeries ohlcPlot(final Comparable seriesName, final Table t, final String time, final String open,
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply Plot/src/main/java/io/deephaven/plot/Axes.java:1064]: no clear edge for either one.

**Comment by jjbrosnan** [comment]
Bug that currently exists in v2 plotting for `plot_cat`:

```
from deephaven2.plot.figure import Figure
from deephaven2.plot import PlotStyle
from deephaven2 import new_table
from deephaven2.column import int_col, string_col

source = new_table([
    string_col("Categories", ["A", "B", "C"]),
    int_col("Values", [9, 10, 11]),
    int_col("LowerError", [-1, -1, -1]),
    int_col("UpperError", [1, 1, 1])
])

figure = Figure()
axes = figure.axes(plot_style = PlotStyle.ERROR_BAR)
new_fig = figure.plot_cat(series_name = "Cagetories Plot", t = source, category = "Categories", y = "Values", y_low = "LowerError", y_high = "UpperError")
new_fig = new_fig.chart_title(title = "Categories and Values")
new_plot = new_fig.show()
```

Attempting to use `y_low` and `y_high` results in a hung plot.  The circle just spins and spins, and the plot remains empty.  The session is still responsive, though.

**Comment by jjbrosnan** [comment]
Plot with error bars in v1:

```
from deephaven import Plot
from deephaven2 import new_table
from deephaven2.column import int_col, string_col

source = new_table([
    string_col("Categories", ["A", "B", "C"]),
    int_col("Categories2", [1, 2, 3]),
    int_col("Values", [9, 10, 11]),
    int_col("LowerError", [8, 9, 10]),
    int_col("UpperError", [10, 11, 12])
])

some_plot = Plot.errorBarY("Categories Plot v1", source.j_table, "Categories2", "Values", "LowerError", "UpperError").show()
```

File: pyintegration/tests/test_plot/test_font.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+        self.assertIsNotNone(custom_font)
+
+    def test_chart_tile_with_font(self):
+        figure = Figure()
+        custom_color = Color.of_hsl(h=128, s=58, l=68, alpha=0.6)
+        custom_font = Font(family="SansSerif", style=FontStyle.BOLD_ITALIC, size=18)
+        new_f = figure.plot_xy("plot1", self.test_table, x="a", y="b")
+        new_chart_title = new_f.chart_title(title="Dummy Char", color=custom_color, font=custom_font)
+        self.assertIsNotNone(new_chart_title)
+
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply pyintegration/tests/test_plot/test_font.py:38]: font constructor doesn't throw exceptions, very error tolerant and just fall back to defaults with bad inputs

**Comment by chipkent** [comment]
The constant spinner for cat error bars is present in v1.
https://github.com/deephaven/deephaven-core/issues/2156

**Comment by chipkent** [comment]
@jjbrosnan y Error bar plotting does seem to be working.

```
from deephaven2.plot.figure import Figure
from deephaven2 import new_table
from deephaven2.column import int_col, string_col

source = new_table([
    string_col("Categories", ["A", "B", "C"]),
    int_col("Categories2", [1, 2, 3]),
    int_col("Values", [9, 10, 11]),
    int_col("LowerError", [8, 9, 10]),
    int_col("UpperError", [10, 11, 12])
])

some_plot = Figure().plot_xy("Error Plot v2", t=source, x="Categories2", y="Values", y_low="LowerError", y_high="UpperError").show()

```

File: pyintegration/deephaven2/plot/selectable_dataset.py [hunk 16]
```diff
@@ -24,3 +29,51 @@ def __init__(self, j_sds: jpy.JType):
...
         return self.j_sds
+
+
+def one_click(t: Table, by: List[str] = None, require_all_filters_to_display: bool = False) -> SelectableDataSet:
+    """ Creates a SelectableDataSet with the specified columns from a table.
+
+    Args:
+        t (Table): the source table
+        by (List[str]): the selected columns
+        require_all_filters_to_display (bool): false to display data when not all oneclicks are selected; true to only
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/selectable_dataset.py:40]: Maybe we could shorten this to `require_all_filters`?

File: pyintegration/deephaven2/plot/selectable_dataset.py [hunk 17]
```diff
@@ -24,3 +29,51 @@ def __init__(self, j_sds: jpy.JType):
...
+    """
+    if not by:
+        by = []
+    try:
+        return SelectableDataSet(j_sds=_JSelectables.oneClick(t.j_table, require_all_filters_to_display, *by))
+    except Exception as e:
+        raise DHError(e, "failed in one_click.") from e
+
+
+def one_click_table_map(tm: jpy.JType, t: Table, by: List[str] = None,
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/selectable_dataset.py:57]: Should we just combine all of these into a `one_click` method?  That seems more pythonic.

File: pyintegration/deephaven2/server/script_session/__init__.py [hunk 18]
```diff
@@ -47,5 +47,7 @@ def unwrap_to_java_type(object):
         return object
     if isinstance(object, JObjectWrapper):
         return object.j_object
+    if isinstance(object, FigureWrapper):
+        return object.figure
```
Comment (chipkent) [inline pyintegration/deephaven2/server/script_session/__init__.py]: I assume we will kill this with v1...

File: pyintegration/tests/test_plot/test_selectable_dataset.py [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+            sds = one_click(self.test_table)
+
+        with self.assertRaises(DHError):
+            sds = one_click(self.test_table)
+
+    def test_one_click_tm(self):
+        tm = self.test_table.j_table.partitionBy("c")
+        sds = one_click_table_map(tm, self.test_table, by=['a', 'b'])
+        self.assertIsNotNone(sds)
+        sds = one_click_table_map(tm, self.test_table, by=['a', 'b'], require_all_filters_to_display=True)
```
Comment (chipkent) [inline pyintegration/tests/test_plot/test_selectable_dataset.py:34]: As a general comment for when we fix things, it seems weird that both `tm` and `t` are needed.  `t` seems unnecessary and like it could be retrieved from `tm`.

File: Generators/Generators.gradle [hunk 20]
```diff
@@ -292,9 +292,38 @@ task generatePythonFigureWrapperAssert(type: JavaExec, dependsOn: 'classes') {
...
+    dependsOn generatePlottingConvenience
+
+    main = 'io.deephaven.plot.util.GeneratePyV2FigureAPI'
+    args devRoot, 'true'
+    classpath = sourceSets.main.runtimeClasspath
+    workingDir = workDir
+    systemProperty 'workspace', workspace
+    systemProperty 'devroot', devRoot
+    systemProperty 'Configuration.rootFile', 'dh-defaults.prop'
+    dependsOn generatePyDoc, ':Generators:generateFigureImmutable'
```
Comment (niloc132) [inline Generators/Generators.gradle:320]: delete

File: Generators/Generators.gradle [hunk 21]
```diff
@@ -292,9 +292,38 @@ task generatePythonFigureWrapperAssert(type: JavaExec, dependsOn: 'classes') {
...
+}
+
+task generatePythonFigureWrapper2Assert(type: JavaExec, dependsOn: 'classes') {
+    description 'Run GeneratePythonFigureWrapper'
+    dependsOn generatePlottingConvenience
```
Comment (niloc132) [inline Generators/Generators.gradle:311]: dependsOn generatePlottingConvenienceAssert

File: pyintegration/deephaven2/plot/__init__.py [hunk 22]
```diff
@@ -0,0 +1,14 @@
+#
+#   Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+""" The plot package includes all the modules for creating plots. """
+
+from .axisformat import AxisFormat, DecimalAxisFormat, NanosAxisFormat
+from .axistransform import AxisTransform, axis_transform_names, axis_transform_by_name
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/__init__.py:7]: Should `axis_transform_by_name` just be `axis_transform`?  This would allow more freedom in the future, if we want to index them other ways.

File: pyintegration/deephaven2/plot/axistransform.py [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+    def __init__(self, j_axis_transform):
+        self.j_axis_transform = j_axis_transform
+
+
+def axis_transform_names() -> List[str]:
+    """ Returns the names of available axis transforms. """
+    return list(_JPlottingConvenience.axisTransformNames())
+
+
+def axis_transform_by_name(name: str) -> AxisTransform:
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/axistransform.py:36]: Rename to `axis_transform` to be concise and allow future indexing in different ways?

File: pyintegration/deephaven2/plot/color.py [hunk 24]
```diff
@@ -0,0 +1,401 @@
+#
+#   Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+""" This module implements the Color class and provides a list of predefined colors that can be used to pain a plot. """
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/color.py:4]:
```suggestion
""" This module implements the Color class and provides a list of predefined colors that can be used to paint a plot. """
```

File: pyintegration/deephaven2/plot/color.py [hunk 25]
```diff
@@ -0,0 +1,401 @@
...
+    TOMATO = Color.of_name("#FF6347")
+    TURQUOISE = Color.of_name("#40E0D0")
+    VIOLET = Color.of_name("#EE82EE")
+    WHEAT = Color.of_name("#F5DEB3")
+    WHITE = Color.of_name("#FFFFFF")
+    WHITESMOKE = Color.of_name("#F5F5F5")
+    YELLOW = Color.of_name("#FFFF00")
+    YELLOWGREEN = Color.of_name("#9ACD32")
+
+    # Legacy non-html-standard colors. Duplicates removed. Conflicts resolved with a DB_ prefix.
```
Comment (chipkent) [inline pyintegration/deephaven2/plot/color.py:269]: \
```suggestion
    # Legacy non-html-standard colors. 
```

File: pyintegration/tests/test_plot/test_font.py [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
```
Comment (chipkent) [inline pyintegration/tests/test_plot/test_font.py:38]: test garbage inputs

**Review by chipkent — CHANGES_REQUESTED** [review]

File: pyintegration/deephaven2/server/script_session/__init__.py (hunk 18 above)
Comment (jmao-denver) [reply pyintegration/deephaven2/server/script_session/__init__.py]: yes when v1 is deleted.

File: pyintegration/tests/test_plot/test_selectable_dataset.py (hunk 19 above)
Comment (jmao-denver) [reply pyintegration/tests/test_plot/test_selectable_dataset.py:34]: what you said makes sense. This must be addressed in the JAVA api first though.

File: pyintegration/deephaven2/plot/selectable_dataset.py (hunk 17 above)
Comment (jmao-denver) [reply pyintegration/deephaven2/plot/selectable_dataset.py:57]: Yeah  definitely doable when the Java API doesn't require both TableMap and Table in the same method.

Comment (jmao-denver) [reply pyintegration/deephaven2/plot/selectable_dataset.py:57]: most likely this will be changed after Ryan completes his rewrite of TableMap.

**Review by chipkent — APPROVED** [review]
